pkg/pillar/nireconciler/genericitems/dnsmasq.go

```diff
@@ -0,0 +1,803 @@
+		return err
+	}
+	defer file.Close()
+	return c.CreateDnsmasqConfig(file, dnsmasq)
+}
+
+// CreateDnsmasqConfig builds configuration for dnsmasq and writes it to the given buffer.
+// The method is exported just to be exercised by unit tests.
+func (c *DnsmasqConfigurator) CreateDnsmasqConfig(buffer io.Writer, dnsmasq Dnsmasq) error {
+	writeErr := func(err error) error {
+		err = fmt.Errorf("failed to write dnsmasq config: %w", err)
+		c.Log.Error(err)
+		return err
+	}
+	if _, err := io.WriteString(buffer, dnsmasqStaticConfig); err != nil {
+		return writeErr(err)
+	}
+	// XXX Have a look at zedrouter's loglevel.
+	// Perhaps we should introduce a separate logger for dnsmasq or per bridge
+	// to have more fine-grained control.
+	switch c.Logger.GetLevel() {
+	case logrus.TraceLevel:
+		if _, err := io.WriteString(buffer, "log-queries\n"); err != nil {
+			return writeErr(err)
+		}
+		if _, err := io.WriteString(buffer, "log-dhcp\n"); err != nil {
+			return writeErr(err)
+		}
+	case logrus.DebugLevel:
+		if _, err := io.WriteString(buffer, "log-dhcp\n"); err != nil {
+			return writeErr(err)
+		}
+	}
+
+	leaseFilepath := devicenetwork.DnsmasqLeaseFilePath(dnsmasq.ListenIf.IfName)
+	if _, err := io.WriteString(buffer,
+		fmt.Sprintf("dhcp-leasefile=%s\n", leaseFilepath)); err != nil {
+		return writeErr(err)
+	}
+
+	// Decide where dnsmasq should send DNS requests upstream.
+	// If we have no uplink for the associated network instance then this is nowhere.
+	// If we have an uplink but no UpstreamServers for it, then we let
+	// dnsmasq use the host's /etc/resolv.conf
+	uplink := dnsmasq.DNSServer.UplinkIf.IfName
+	if uplink == "" {
+		if _, err := io.WriteString(buffer, "no-resolv\n"); err != nil {
+			return writeErr(err)
+		}
+	} else if len(dnsmasq.DNSServer.UpstreamServers) != 0 {
+		for _, s := range dnsmasq.DNSServer.UpstreamServers {
+			_, err := io.WriteString(buffer, fmt.Sprintf("server=%s@%s\n", s, uplink))
+			if err != nil {
+				return writeErr(err)
+			}
+		}
+		if _, err := io.WriteString(buffer, "no-resolv\n"); err != nil {
+			return writeErr(err)
+		}
+	}
+
+	for _, ipset := range dnsmasq.DNSServer.LinuxIPSets {
+		if _, err := io.WriteString(buffer,
+			fmt.Sprintf("ipset=/%s/%s\n",
+				strings.Join(ipset.Domains, "/"),
+				strings.Join(ipset.Sets, ","))); err != nil {
+			return writeErr(err)
+		}
+	}
+
+	pidFile := c.dnsmasqPidFile(dnsmasq.InstanceName)
+	if _, err := io.WriteString(buffer,
+		fmt.Sprintf("pid-file=%s\n", pidFile)); err != nil {
+		return writeErr(err)
+	}
+
+	if _, err := io.WriteString(buffer,
+		fmt.Sprintf("interface=%s\n", dnsmasq.ListenIf.IfName)); err != nil {
+		return writeErr(err)
+	}
+	isIPv6 := false
+	listenIP := dnsmasq.DNSServer.ListenIP
+	if listenIP != nil {
+		isIPv6 = listenIP.To4() == nil
+		if _, err := io.WriteString(buffer,
+			fmt.Sprintf("listen-address=%s\n", listenIP)); err != nil {
+			return writeErr(err)
+		}
+	} else {
+		// XXX Error if there is no ListenIP?
```

Comment:
Should we maybe be explicit here? I.e. if listen address is any we specify it as 0.0.0.0 and then we won't have this problem.